rust/processor/src/bq_analytics/generic_parquet_processor.rs

```diff
@@ -153,84 +161,106 @@ where
-            .await;
-            self.buffer_size_bytes = 0;
-            remove_file(&new_file_path)?;
-
-            return match upload_result {
-                Ok(_) => {
-                    let parquet_processing_result = ParquetProcessingResult {
-                        start_version,
-                        end_version,
-                        last_transaction_timestamp: Some(last_transaction_timestamp),
-                        txn_version_to_struct_count,
-                    };
-
-                    self.gap_detector_sender
-                        .send(ProcessingResult::ParquetProcessingResult(
-                            parquet_processing_result,
-                        ))
-                        .await
-                        .expect("[Parser] Failed to send versions to gap detector");
-                    Ok(())
-                },
-                Err(e) => {
-                    error!("Failed to upload file to GCS: {}", e);
-                    Err(anyhow!("Failed to upload file to GCS: {}", e))
-                },
-            };
+            if self.buffer_size_bytes >= self.max_buffer_size {
+                info!("Max buffer size reached, uploading to GCS.");
+                if let Err(e) = self.upload_buffer(gcs_client).await {
+                    error!("Failed to upload buffer: {}", e);
+                    return Err(e);
+                }
+                self.last_upload_time = Instant::now();
+            }
+
+            if self.last_upload_time.elapsed() >= self.upload_interval {
+                info!(
+                    "Time has elapsed more than {} since last upload.",
+                    self.upload_interval.as_secs()
+                );
+                if let Err(e) = self.upload_buffer(gcs_client).await {
+                    error!("Failed to upload buffer: {}", e);
+                    return Err(e);
+                }
+                self.last_upload_time = Instant::now();
+            }
         }
 
         PARQUET_HANDLER_BUFFER_SIZE
             .with_label_values(&[ParquetType::TABLE_NAME])
             .set(self.buffer.len() as i64);
         Ok(())
     }
+
+    pub async fn upload_buffer(&mut self, gcs_client: &GCSClient) -> Result<()> {
+        if self.buffer.is_empty() {
+            return Ok(());
+        }
+        let start_version = self.buffer.first().unwrap().version();
+        let last = self.buffer.last().unwrap();
+        let end_version = last.version();
+        let last_transaction_timestamp = naive_datetime_to_timestamp(last.get_timestamp());
+
+        let txn_version_to_struct_count =
+            process_struct_count_map(&self.buffer, &mut self.transaction_version_to_struct_count);
+
+        let new_file_path: PathBuf = PathBuf::from(format!(
+            "{}_{}.parquet",
+            ParquetType::TABLE_NAME,
+            Uuid::new_v4()
+        ));
+        rename(&self.file_path, &new_file_path)?; // this fixes an issue with concurrent file access issues
```

Comment:
nit: document more around the file rename... what's the issue here btw?

Reply:
it's been a while.. but I recall that there is a issue with file not found or something like after first upload. I can revisit this later and see if we can remove this behavior

Reply:
can you elaborate a little? do you mean file not found right when reading into bigquery right after upload?